### ISSUE TITLE
nullable:true doesn't accept null value

### ISSUE THREAD

**Issue by SlytherinCz** [oas-tools/oas-tools#164]
oas-tools version 2.1.6
openapi: 3.0.0

**Schema**
```
...
company:
  type: object
  properties:
    taxDomicile:
      nullable: true
      type: string
      minLength: 3
      maxLength: 3
...
```

**Request**
```
...
"company": {
        "taxDomicile": null
    }
...
```

**Message during validation**
```
...{"code":"INVALID_TYPE","params":["string","null"],"message":"Expected type string but found type null","path":"#/comp
any/taxDomicile"}...
```

I expected that `nullable` would allow me to send `null`. I can only work around this by not sending the field at all. That approach would mean I would have to sanitize each outgoing message on client, which is something i would very happily avoid.

Is this expected behavior?

**Comment by SlytherinCz** [comment]
Some further details to consider:

The error is originating from z-schema package, in file `src/schemas/JsonValidation.js:268` , method `type` which doesn't take nullable under consideration.

Code in question:
```
type: function (report, schema, json) {
        // http://json-schema.org/latest/json-schema-validation.html#rfc.section.5.5.2.2
        var jsonType = Utils.whatIs(json);        
        if (typeof schema.type === "string") {
            if (jsonType !== schema.type && (jsonType !== "integer" || schema.type !== "number")) {
                report.addError("INVALID_TYPE", [schema.type, jsonType], null, schema);
            }
        } else {
            if (schema.type.indexOf(jsonType) === -1 && (jsonType !== "integer" || schema.type.indexOf("number") === -1)) {
                report.addError("INVALID_TYPE", [schema.type, jsonType], null, schema);
            }
        }
    }
```

The validation follows somewhat the following process:
1) Automatically invalid if schema is not an object
2) Automatically valid if schema is empty
3) combine references to create a complete schema
4) validate types

During the step 4, aforementioned method `type` is called, generating error (probably correctly, because `null` indeed is not a `string`)

From my perspective, the `nullable` part is not a concern of z-schema, rather a feature of OpenApi, therefor I would think these errors should be discarded on the side of oas-tools in case of nullable field.

Are we on the same page, so to speak?
Also, is there something I can do to help implement this?

**Comment by SlytherinCz** [comment]
Even more things to consider: 

The following code handles the described scenario and discards error originating from z-schema, even though values might not be erroneous from our point of view. This code replaces the closing of function `checkRequestData` in file `src/middleware/oas-validator.js:126` and onwards.

```
if (err == false) {
  var originalErrors = validator.getLastErrors()
  var keptErrors = []
  for(let errorIndex in originalErrors) {
    if (
      !( originalErrors[errorIndex].code === 'INVALID_TYPE'
        && originalErrors[errorIndex].params[1] === 'null'
        && originalErrors[errorIndex][Symbol.for("z-schema/schema")].nullable
      )
    ) {
      keptErrors.push(originalErrors[errorIndex])
    }
  }
  if(keptErrors.length) {
    newErr = {
      message: "Wrong data in the body of the request. ",
      error: keptErrors,
      content: data
    };
    msg.push(newErr);
    keepGoing = false;
  } else {
    logger.info("Valid parameter on request");
  }
} else {
  logger.info("Valid parameter on request");
}
```

This is not unit tested and I only tried it against my particular schema and dataset. Would you consider implementing some similar approach? 
Or even, do you so far agree with my general thought process regarding this specific use case?

Questions to answer are:
Is this intended behavior?
Do you agree this should be handled on oas-tools side?
Is discarding (filtering) error messages correct approach to this?

**Comment by yitzgold** [comment]
This issue has been fixed on the master branch it just hasn't been deployed yet

**Comment by SlytherinCz** [comment]
It seems to me that fix only handles strings. What about all the other cases?

**Comment by raffrearaUS** [comment]
We have just released a new version containing the fix that @yitzgold mentioned. Could you please check if it is working for you now?

**Comment by alesancor1** [comment]
A new release that solves this issue has been published (`v3.0.2`). Please run `npm update @oas-tools/core` to update OAS Tools core library.
